README.md
# RTL8720_RTC Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/RTL8720_RTC.svg?)](https://www.ardu-badge.com/RTL8720_RTC)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/RTL8720_RTC.svg)](https://github.com/khoih-prog/RTL8720_RTC/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/RTL8720_RTC/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/RTL8720_RTC.svg)](http://github.com/khoih-prog/RTL8720_RTC/issues)

<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents

* [Why do we need this RTL8720_RTC library](#why-do-we-need-this-rtl8720_rtc-library)
  * [Features](#features)
  * [Why using ISR-based Alarm is better](#why-using-isr-based-alarm-is-better)
  * [Currently supported Boards](#currently-supported-boards)
  * [Important Notes about ISR](#important-notes-about-isr)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Packages' Patches](#packages-patches)
  * [1. For RTL8720DN boards using AmebaD core](#1-for-rtl8720dn-boards-using-amebad-core)
* [Usage](#usage)
  * [1. RTL8720_RTC Class](#1-rtl8720_rtc-class)
    * [1.1 Function List](#11-function-list)
  * [2. TimeSpan and DateTime](#2-timespan-and-datetime)
    * [2.1 Typedef](#21-typedef)
    * [2.2 Function List](#22-function-list)
* [Examples](#examples)
  * [ 1. RTL8720_RTC_Time_WiFi](examples/RTL8720_RTC_Time_WiFi)
* [Example RTL8720_RTC_Time_WiFi](#example-rtl8720_rtc_time_wifi)
  * [ 1. File RTL8720_RTC_Time_WiFi.ino](#1-file-rtl8720_rtc_time_wifiino)
  * [ 2. File defines.h](#2-file-definesh)
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  * [1. RTL8720_RTC_Time_WiFi on RTL8720DN_BW16](#1-rtl8720_rtc_time_wifi-on-rtl8720dn_bw16)
* [Troubleshooting](#troubleshooting)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License](#license)
* [Copyright](#copyright)

---
---

### Why do we need this [RTL8720_RTC library](https://github.com/khoih-prog/RTL8720_RTC)

## Features

This is an Arduino library for **RTL8720-based RTC**

The examples will demonstrate how to get the UTC time from NTP server, then update the RTC to make sure the time is perfectly correct.


### Why using ISR-based Alarm is better

Imagine you have a system with a **mission-critical** function, measuring water level and control the sump pump or doing something much more important. You normally use a software timer to poll, or even place the function in loop(). But what if another function is **blocking** the loop() or setup().

So your function **might not be executed, and the result would be disastrous.**

You'd prefer to have your function called, no matter what happening with other functions (busy loop, bug, etc.).

The correct choice is to use an ISR-based Alarm with **Interrupt** to call your function.

These ISR-based Alarm, using interrupt, still work even if other functions are blocking. Moreover, they are much more **precise** (certainly depending on clock frequency accuracy) than other software timers using millis() or micros(). That's necessary if you need to measure some data requiring better accuracy.

Functions using normal software timers, relying on loop() and calling millis(), won't work if the loop() or setup() is blocked by certain operation. For example, certain function is blocking while it's connecting to WiFi or some services.

The catch is **your function is now part of an ISR (Interrupt Service Routine), and must be lean / mean, and follow certain rules.** More to read on:

[**HOWTO Attach Interrupt**](https://www.arduino.cc/reference/en/language/functions/external-interrupts/attachinterrupt/)

---

### Currently supported Boards

1. RTL8720-based boards such as **Realtek RTL8720DN, RTL8722DM and RTL8722CSM**, etc.

---

### Important Notes about ISR

1. Inside the attached function, **delay() won’t work and the value returned by millis() will not increment.** Serial data received while in the function may be lost. You should declare as **volatile any variables that you modify within the attached function.**

2. Typically global variables are used to pass data between an ISR and the main program. To make sure variables shared between an ISR and the main program are updated correctly, declare them as volatile.

3. Don't use Serial.print(ln) function inside ISR or system will hang up.

---
---

## Prerequisites

1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
2. [`Arduino AmebaD core 3.1.2+`](https://github.com/ambiot/ambd_arduino) for Realtek RTL8720DN, RTL8722DM and RTL8722CSM. [![GitHub release](https://img.shields.io/github/release/ambiot/ambd_arduino.svg)](https://github.com/ambiot/ambd_arduino/releases/latest)
3. [`Timezone_Generic library v1.10.0+`](https://github.com/khoih-prog/Timezone_Generic) to use examples using Timezone. To install, check [![arduino-library-badge](https://www.ardu-badge.com/badge/Timezone_Generic.svg?)](https://www.ardu-badge.com/Timezone_Generic)
4. [`WiFiWebServer_RTL8720 library v1.1.1+`](https://github.com/khoih-prog/WiFiWebServer_RTL8720) WiFi to access NTP server. To install. check [![arduino-library-badge](https://www.ardu-badge.com/badge/WiFiWebServer_RTL8720.svg?)](https://www.ardu-badge.com/WiFiWebServer_RTL8720).

---
---

## Installation

### Use Arduino Library Manager

The best and easiest way is to use `Arduino Library Manager`. Search for [**RTL8720_RTC**](https://github.com/khoih-prog/RTL8720_RTC), then select / install the latest version.
You can also use this link [![arduino-library-badge](https://www.ardu-badge.com/badge/RTL8720_RTC.svg?)](https://www.ardu-badge.com/RTL8720_RTC) for more detailed instructions.

### Manual Install

Another way to install is to:

1. Navigate to [**RTL8720_RTC**](https://github.com/khoih-prog/RTL8720_RTC) page.
2. Download the latest release `RTL8720_RTC-main.zip`.
3. Extract the zip file to `RTL8720_RTC-main` directory 
4. Copy whole `RTL8720_RTC-main` folder to Arduino libraries' directory such as `~/Arduino/libraries/`.

### VS Code & PlatformIO

1. Install [VS Code](https://code.visualstudio.com/)
2. Install [PlatformIO](https://platformio.org/platformio-ide)
3. Install [**RTL8720_RTC** library](https://registry.platformio.org/libraries/khoih-prog/RTL8720_RTC) by using [Library Manager](https://registry.platformio.org/libraries/khoih-prog/RTL8720_RTC/installation). Search for **RTL8720_RTC** in [Platform.io Author's Libraries](https://platformio.org/lib/search?query=author:%22Khoi%20Hoang%22)
4. Please visit documentation for the other options and examples at [Project Configuration File](https://docs.platformio.org/page/projectconf.html)


---
---

### Packages' Patches

#### 1. For RTL8720DN boards using AmebaD core
 
 To avoid compile error relating to PROGMEM, you have to copy the file [Realtek AmebaD core pgmspace.h](Packages_Patches/realtek/hardware/AmebaD/3.1.1/cores/arduino/avr/pgmspace.h) into Realtek AmebaD directory (~/.arduino15/packages/realtek/hardware/AmebaD/3.1.1/cores/arduino/avr/pgmspace.h). 

Supposing the Realtek AmebaD core version is 3.1.1. This file must be copied into the directory:

- `~/.arduino15/packages/realtek/hardware/AmebaD/3.1.1/cores/arduino/avr/pgmspace.h`

Whenever a new version is installed, remember to copy this file into the new version directory. For example, new version is x.yy.zz
This file must be copied into the directory:

- `~/.arduino15/packages/realtek/hardware/AmebaD/x.yy.zz/cores/arduino/avr/pgmspace.h`

---
---

## Usage

### 1. RTL8720_RTC Class

### 1.1. Function List

```
void init();
bool free();
void write(time_t t);
time_t read();
void wait(float s);
void setAlarm(alarm_t *alrm, alarm_irq_handler alarmHandler);
void disableAlarm();
void restoreTimeInfo();
void backupTimeInfo();
```

---

### 2. TimeSpan and DateTime

### 2.1. Typedef

```
class TimeSpan;

class DateTime;

/** ISO 8601 Timestamp function */
enum timestampOpt 
{
  TIMESTAMP_FULL, // YYYY-MM-DDTHH:MM:SS
  TIMESTAMP_TIME, // HH:MM:SS
  TIMESTAMP_DATE  // YYYY-MM-DD
};
```

### 2.2. Function List

```
class DateTime
{
  DateTime (uint32_t t = SECONDS_FROM_1970_TO_2000);
  DateTime (const uint16_t year, const uint8_t month, const uint8_t day, const uint8_t hour = 0, const uint8_t min = 0, const uint8_t sec = 0);
  DateTime (const DateTime& copy);
  DateTime (const tmElements_t& tm);

  // To use with RTL8720 datetime_t struct
  DateTime (const datetime_t &tm);

  tmElements_t get_tmElements_t();
  void setFrom_tmElements_t(const tmElements_t& tm);
  time_t get_time_t();
  void setFrom_time_t(const time_t& timeInput);

  uint16_t year() const;
  uint8_t month() const;
  uint8_t day() const;
  uint8_t hour() const;
  uint8_t minute() const;
  uint8_t second() const;
  uint16_t yearOffset() const;
  uint8_t dayOfTheWeek() const;

  /** 32-bit times as seconds since 1/1/2000 */
  long secondstime() const;

  /** 32-bit times as seconds since 1/1/1970 */
  uint32_t unixtime() const;

  /** ISO 8601 Timestamp function */
  String timestamp(timestampOpt opt = TIMESTAMP_FULL)
}

```


---
---

### Examples 

 1. [RTL8720_RTC_Time_WiFi](examples/RTL8720_RTC_Time_WiFi)

---
---

### Example [RTL8720_RTC_Time_WiFi](examples/RTL8720_RTC_Time_WiFi)

#### 1. File [RTL8720_RTC_Time_WiFi.ino](examples/RTL8720_RTC_Time_WiFi/RTL8720_RTC_Time_WiFi.ino)


https://github.com/khoih-prog/RTL8720_RTC/blob/92386b3e49bd8028ce108aac46a3fb13bc87fe3e/examples/RTL8720_RTC_Time_WiFi/RTL8720_RTC_Time_WiFi.ino#L15-L260



#### 2. File [defines.h](examples/RTL8720_RTC_Time_WiFi/defines.h)

https://github.com/khoih-prog/RTL8720_RTC/blob/92386b3e49bd8028ce108aac46a3fb13bc87fe3e/examples/RTL8720_RTC_Time_WiFi/defines.h#L15-L29

---
---

### Debug Terminal Output Samples

#### 1. RTL8720_RTC_Time_WiFi on RTL8720DN_BW16

The following is the sample terminal output when running example [RTL8720_RTC_Time_WiFi](examples/RTL8720_RTC_Time_WiFi) on RTL8720DN_BW16

```
Start RTL8720_RTC_Time_WiFi on RTL8720DN_BW16
RTL8720_RTC v1.0.2
Timezone_Generic v1.10.0
Current Firmware Version = 1.0.0
Attempting to connect to SSID: HueNet_5G
Attempting to connect to SSID: HueNet_5G
TZ_NTP_Clock_RTL8720DN started @ IP address: 192.168.2.111
Packet received
Seconds since Jan 1 1900 = 3851712462
Unix time = 1642723662
The UTC time is 0:07:42
============================
00:07:43 Fri 21 Jan 2022 UTC
19:07:43 Thu 20 Jan 2022 EST
```

---
---

### Troubleshooting

If you get compilation errors, more often than not, you may need to install a newer version of the core for Arduino boards.

Sometimes, the library will only work if you update the board core to the latest version because I am using newly added functions.


---
---

### Issues

Submit issues to: [RTL8720_RTC issues](https://github.com/khoih-prog/RTL8720_RTC/issues)

---

## TO DO

1. Search for bug and improvement.


---

## DONE

1. Basic RTC-wrapper library for **Realtek RTL8720DN, RTL8722DM and RTL8722CSM**, using [`Arduino AmebaD core`](https://github.com/ambiot/ambd_arduino)
2. Add Version String 
3. Add Table of Contents

---
---

### Contributions and Thanks

Many thanks for everyone for bug reporting, new feature suggesting, testing and contributing to the development of this library.

---

## Contributing

If you want to contribute to this project:

- Report bugs and errors
- Ask for enhancements
- Create issues and pull requests
- Tell other people about this library

---

### License

- The library is licensed under [MIT](https://github.com/khoih-prog/RTL8720_RTC/blob/main/LICENSE)

---

## Copyright

Copyright 2021- Khoi Hoang




changelog.md
# RTL8720_RTC Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/RTL8720_RTC.svg?)](https://www.ardu-badge.com/RTL8720_RTC)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/RTL8720_RTC.svg)](https://github.com/khoih-prog/RTL8720_RTC/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/RTL8720_RTC/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/RTL8720_RTC.svg)](http://github.com/khoih-prog/RTL8720_RTC/issues)

---
---

## Table of Contents

* [Changelog](#changelog)
  * [Releases v1.0.2](#releases-v102)
  * [Releases v1.0.1](#releases-v101)
  * [Initial Releases v1.0.0](#initial-releases-v100)

---
---

## Changelog

### Releases v1.0.2

1. Update dependency for PlatformIO
2. Update `Packages_Patches`

### Releases v1.0.1

1. Update dependency on [Timezone_Generic Library](https://github.com/khoih-prog/Timezone_Generic)
2. Update examples
3. Update `Packages_Patches`

### Initial Releases v1.0.0

1. Initial coding to support RTL8720-based boards such as **Realtek RTL8720DN, RTL8722DM and RTL8722CSM** using [`Arduino AmebaD core`](https://github.com/ambiot/ambd_arduino)




src/RTL8720_RTC.h
/****************************************************************************************************************************
  RTL8720_RTC.h
  For RTL8720-based boards using internal RTC with RTL8720DN WiFi

  Built by Khoi Hoang https://github.com/khoih-prog/RTL8720_RTC
  Licensed under MIT license
  
  Version: 1.0.2

  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0  K Hoang      10/08/2021 Initial release to support RTL8720-based boards using internal RTC with RTL8720DN WiFi
  1.0.1  K Hoang      20/01/2022 Update dependency
  1.0.2  K Hoang      27/04/2022 Update dependency for PlatformIO
*****************************************************************************************************************************/

#pragma once

#ifndef RTL8720_RTC_H
#define RTL8720_RTC_H

#if ( defined(CONFIG_PLATFORM_8721D) || defined(BOARD_RTL8722D) || defined(BOARD_RTL8722DM_MINI) || defined(BOARD_RTL8720DN_BW16) )

  #if defined(BOARD_RTL8722D)
    #define BOARD_NAME      "RTL8722D"
  #elif defined(BOARD_RTL8722DM_MINI)
    #define BOARD_NAME      "RTL8722DM_MINI"
  #elif defined(BOARD_RTL8720DN_BW16)
    #define BOARD_NAME      "RTL8720DN_BW16"
  #endif

#else
  #error This code is designed to run on RTL8720-based platform! Please check your Tools->Board setting.
#endif

#define RTL8720_RTC_VERSION             "RTL8720_RTC v1.0.2"

#define RTL8720_RTC_VERSION_MAJOR       1
#define RTL8720_RTC_VERSION_MINOR       0
#define RTL8720_RTC_VERSION_PATCH       2

#define RTL8720_RTC_VERSION_INT         1000002

#include <Arduino.h>

#include "rtc_api.h"
#include "timer_api.h"

#ifdef __cplusplus
  extern "C" {
#endif

  /*static*/ void rtc_restore_timeinfo();
  void rtc_backup_timeinfo();
  
#ifdef __cplusplus
  }
#endif

class RTL8720_RTC
{
  public:

    RTL8720_RTC() {};
    ~RTL8720_RTC() {};

    /**
        @brief  Initializes the RTC device, include clock, RTC registers and function.
        @param  none
        @retval  none
    */
    void init() 
    {
      rtc_init();
    }

    /**
        @brief  Deinitializes the RTC device.
        @param  none
        @retval  none
    */
    void free() 
    {
      rtc_free();
    }

    /**
        @brief  Set the specified timestamp in seconds to RTC.
        @param  t: Seconds from 1970.1.1 00:00:00 to specified data and time
                     which is to be set.
        @retval  none
    */
    void write(time_t t)
    {
      rtc_write( (time_t) t);  
    }

    /**
        @brief  Get current timestamp in seconds from RTC.
        @param  none
        @retval value: The current timestamp in seconds which is calculated from 1970.1.1 00:00:00.
    */
    time_t read()
    {
      return (rtc_read());
    }

    /**
        @brief  Wait for 1000000 us / 1s.
    */
    void wait(float s) 
    {
      DelayUs((int)(s * 1000000.0f));
    }

    /**
      * @brief  Set the specified RTC Alarm and interrupt.
      * @param  alarm: alarm object define in application software.
      * @param  alarmHandler:  alarm interrupt callback function.
      * @retval   status:
      */
      
    void setAlarm(alarm_t *alrm, alarm_irq_handler alarmHandler)
    {
	    rtc_set_alarm(alrm, alarmHandler);
    }  

    /**
      * @brief  Disable RTC Alarm and function.
      * @param  none
      * @retval   none
      */
    void disableAlarm()
    {
	    rtc_disable_alarm();
    }

    /**
      * @brief  This function is used to restore rtc_timeinfo global variable whose value is lost after system reset.
      * @param  none
      * @retval value: none
      */    
    void restoreTimeInfo()
    {
      rtc_restore_timeinfo();
    }
    
    /**
      * @brief  This function is used to backup tm_year parameter in rtc_timeinfo global variable before system reset.
      * @param  none
      * @retval value: none
      */
    void backupTimeInfo()
    {
	    rtc_backup_timeinfo();
    }

  private:
  

};

#include "DateTime_Generic.h"

#endif    // RTL8720_RTC_H


examples/RTL8720_RTC_Time_WiFi/defines.h
/****************************************************************************************************************************
  defines.h

  For RTL8720-based boards using internal RTC with RTL8720DN WiFi

  Built by Khoi Hoang https://github.com/khoih-prog/RTL8720_RTC
  Licensed under MIT license
  Version: 1.0.0

  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0  K Hoang      10/08/2021 Initial release to support RTL8720-based boards using internal RTC with RTL8720DN WiFi
 *****************************************************************************************************************************/

#ifndef defines_h
#define defines_h

#if !( defined(CONFIG_PLATFORM_8721D) || defined(BOARD_RTL8722D) || defined(BOARD_RTL8722DM_MINI) || defined(BOARD_RTL8720DN_BW16) )
  #error This code is designed to run on RTL8720-based platform! Please check your Tools->Board setting.
#endif

#include <RTL8720_RTC.h>

#include <WiFiWebServer_RTL8720.h>

char ssid[] = "SSID_5GHz";        // your network SSID (name)
char pass[] = "12345678";        // your network password

#endif    //defines_h


src/DateTime_Generic.h
/****************************************************************************************************************************
  DateTime_Generic.h
  
  For RTL8720-based boards using internal RTC with RTL8720DN WiFi

  Built by Khoi Hoang https://github.com/khoih-prog/RTL8720_RTC
  Licensed under MIT license
  
  Version: 1.0.2

  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0  K Hoang      10/08/2021 Initial release to support RTL8720-based boards using internal RTC with RTL8720DN WiFi
  1.0.1  K Hoang      20/01/2022 Update dependency
  1.0.2  K Hoang      27/04/2022 Update dependency for PlatformIO
 *****************************************************************************************************************************/
 
#pragma once

#ifndef DATETIME_GENERIC_H
#define DATETIME_GENERIC_H

#include <TimeLib.h>    // https://github.com/PaulStoffregen/Time

// port of https://github.com/adafruit/RTClib

class TimeSpan
{
  protected:

    int32_t _seconds;   ///< Actual TimeSpan value is stored as seconds

  public:

    TimeSpan (const int32_t seconds = 0)
      : _seconds(seconds)
    {}
    
    TimeSpan (const int16_t days, const int8_t hours, const int8_t minutes, const int8_t seconds)
      : _seconds((int32_t)days * 86400L + (int32_t)hours * 3600 + (int32_t)minutes * 60 + seconds)
    {}
    
    TimeSpan (const TimeSpan& copy)
      : _seconds(copy._seconds)
    {}

    int16_t days() const
    {
      return _seconds / 86400L;
    }
    
    int8_t  hours() const
    {
      return _seconds / 3600 % 24;
    }
    
    int8_t  minutes() const
    {
      return _seconds / 60 % 60;
    }
    
    int8_t  seconds() const
    {
      return _seconds % 60;
    }
    
    int32_t totalseconds() const
    {
      return _seconds;
    }

    TimeSpan operator+(const TimeSpan& right)
    {
      return TimeSpan(_seconds + right._seconds);
    }
    
    TimeSpan operator-(const TimeSpan& right)
    {
      return TimeSpan(_seconds - right._seconds);
    }
};


class DateTime
{
  protected:

    uint8_t yOff;   ///< Year offset from 2000
    uint8_t m;      ///< Month 1-12
    uint8_t d;      ///< Day 1-31
    uint8_t hh;     ///< Hours 0-23
    uint8_t mm;     ///< Minutes 0-59
    uint8_t ss;     ///< Seconds 0-59

    static constexpr uint32_t SECONDS_FROM_1970_TO_2000 {946684800};
    const uint8_t daysInMonth[12] {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};

  public:

    // Can use DateTime(epoch) where epoch is Unix time
    DateTime (uint32_t t = SECONDS_FROM_1970_TO_2000)
    {
      t -= SECONDS_FROM_1970_TO_2000;    // bring to 2000 timestamp from 1970

      ss = t % 60;
      t /= 60;
      
      mm = t % 60;
      t /= 60;
      
      hh = t % 24;
      uint16_t days = t / 24;
      uint8_t leap;
      
      for (yOff = 0; ; ++yOff) 
      {
        leap = yOff % 4 == 0;
        
        if ( days < (uint16_t) (365 + leap) )
          break;
          
        days -= 365 + leap;
      }
      
      for (m = 1; ; ++m) 
      {
        uint8_t daysPerMonth = daysInMonth[m - 1];
        
        if (leap && m == 2)
          ++daysPerMonth;
          
        if (days < daysPerMonth)
          break;
          
        days -= daysPerMonth;
      }
      
      d = days + 1;
    }

    DateTime (const uint16_t year, const uint8_t month, const uint8_t day, const uint8_t hour = 0, const uint8_t min = 0, const uint8_t sec = 0)
    {
      if (year >= 2000)
        yOff = year - 2000;
      else
        yOff = year;
        
      m   = month;
      d   = day;
      hh  = hour;
      mm  = min;
      ss  = sec;
    }

    DateTime (const DateTime& copy)
      : yOff(copy.yOff) , m(copy.m) , d(copy.d) , hh(copy.hh), mm(copy.mm), ss(copy.ss)
    {}

    // DateTime (const char* date, const char* time)
    // {
    //     // sample input: date = "Dec 26 2009", time = "12:34:56"
    //     yOff = conv2d(date + 9);
    //     // Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec
    //     switch (date[0]) {
    //         case 'J': m = (date[1] == 'a') ? 1 : ((date[2] == 'n') ? 6 : 7); break;
    //         case 'F': m = 2; break;
    //         case 'A': m = date[2] == 'r' ? 4 : 8; break;
    //         case 'M': m = date[2] == 'r' ? 3 : 5; break;
    //         case 'S': m = 9; break;
    //         case 'O': m = 10; break;
    //         case 'N': m = 11; break;
    //         case 'D': m = 12; break;
    //     }
    //     d = conv2d(date + 4);
    //     hh = conv2d(time);
    //     mm = conv2d(time + 3);
    //     ss = conv2d(time + 6);
    // }

    // DateTime (const __FlashStringHelper* date, const __FlashStringHelper* time);
    // char* toString(char* buffer);
    
    //////////////////////////////////
    // KH add
    
    #define YEARS_FROM_1970_TO_2000     30
     
    DateTime (const tmElements_t& tm)
    {     
      setFrom_tmElements_t(tm);
    }
    
    DateTime (const time_t& timeInput)
    {
      setFrom_time_t(timeInput);
    }
 
    tmElements_t get_tmElements_t()
    {
      tmElements_t tm;

      // tmElements_t year counts from 1970
      tm.Year     = this->year() - 1970;
      tm.Month    = this->month();
      tm.Day      = this->day();
      tm.Hour     = this->hour();
      tm.Minute   = this->minute();
      tm.Second   = this->second(); 
      
      return tm;   
    }
    
    void setFrom_tmElements_t(const tmElements_t& tm)
    {
      // tm.Year from 1970
      if (tm.Year >= YEARS_FROM_1970_TO_2000)
        yOff = tm.Year - YEARS_FROM_1970_TO_2000;
      else
        yOff = tm.Year + 1970;
        
      m   = tm.Month;
      d   = tm.Day;
      hh  = tm.Hour;
      mm  = tm.Minute;
      ss  = tm.Second;
    }
    
    time_t get_time_t()
    {
      return makeTime(get_tmElements_t());
    }   

    void setFrom_time_t(const time_t& timeInput)
    {
      tmElements_t tm;
      
      breakTime(timeInput, tm);
  
      // tm.Year from 1970
      if (tm.Year >= YEARS_FROM_1970_TO_2000)
        yOff = tm.Year - YEARS_FROM_1970_TO_2000;
      else
        yOff = tm.Year + 1970;
        
      m   = tm.Month;
      d   = tm.Day;
      hh  = tm.Hour;
      mm  = tm.Minute;
      ss  = tm.Second;
    }   
    
    //////////////////////////////////

    uint16_t year() const 
    {
      return 2000 + yOff;
    }
    
    uint8_t month() const 
    {
      return m;
    }
    
    uint8_t day() const 
    {
      return d;
    }
    
    uint8_t hour() const 
    {
      return hh;
    }
    
    uint8_t minute() const 
    {
      return mm;
    }
    
    uint8_t second() const 
    {
      return ss;
    }

    uint16_t yearOffset() const 
    {
      return yOff;
    }

    uint8_t dayOfTheWeek() const
    {
      uint16_t day = date2days(yOff, m, d);
      
      return (day + 6) % 7; // Jan 1, 2000 is a Saturday, i.e. returns 6
    }

    /** 32-bit times as seconds since 1/1/2000 */
    long secondstime() const
    {
      long t;
      uint16_t days = date2days(yOff, m, d);
      
      t = time2long(days, hh, mm, ss);
      
      return t;
    }

    /** 32-bit times as seconds since 1/1/1970 */
    uint32_t unixtime() const
    {
      uint32_t t;
      uint16_t days = date2days(yOff, m, d);
      
      t = time2long(days, hh, mm, ss);
      t += SECONDS_FROM_1970_TO_2000;  // seconds from 1970 to 2000

      return t;
    }

    /** ISO 8601 Timestamp function */
    enum timestampOpt {
      TIMESTAMP_FULL, // YYYY-MM-DDTHH:MM:SS
      TIMESTAMP_TIME, // HH:MM:SS
      TIMESTAMP_DATE  // YYYY-MM-DD
    };

    String timestamp(timestampOpt opt = TIMESTAMP_FULL)
    {
      char buffer[28];

      // Generate timestamp according to opt
      switch (opt) 
      {
        case TIMESTAMP_TIME:
          // Only time
          sprintf(buffer, "%02d:%02d:%02d", hh, mm, ss);
          break;
        case TIMESTAMP_DATE:
          // Only date
          sprintf(buffer, "%04d-%02d-%02d", 2000 + yOff, m, d);
          break;
        default:
          // Full
          sprintf(buffer, "%04d-%02d-%02dT%02d:%02d:%02d", 2000 + yOff, m, d, hh, mm, ss);
      }
      
      return String(buffer);
    }

    DateTime operator+(const TimeSpan& span)
    {
      return DateTime( ( unixtime() + (uint32_t) span.totalseconds() ));
    }
    
    DateTime operator-(const TimeSpan& span)
    {
      return DateTime( ( unixtime() - (uint32_t) span.totalseconds() ));
    }
    
    TimeSpan operator-(const DateTime& right)
    {
      return TimeSpan( (int32_t) ( unixtime() - right.unixtime() ));
    }
    
    bool operator<(const DateTime& right) const
    {
      return unixtime() < right.unixtime();
    }
    
    bool operator>(const DateTime& right) const
    {
      return right < *this;
    }
    
    bool operator<=(const DateTime& right) const
    {
      return !(*this > right);
    }
    
    bool operator>=(const DateTime& right) const
    {
      return !(*this < right);
    }
    
    bool operator==(const DateTime& right) const
    {
      return unixtime() == right.unixtime();
    }
    
    bool operator!=(const DateTime& right) const
    {
      return !(*this == right);
    }

  protected:

    uint16_t date2days(const uint16_t year, const uint8_t m, const uint8_t d) const
    {
      uint16_t y;
      
      if (year >= 2000) 
        y = year - 2000;
      else
        y = year;
        
      uint16_t days = d;
      
      for (uint8_t i = 1; i < m; ++i)
        days += daysInMonth[i - 1];
        
      if (m > 2 && y % 4 == 0)
        ++days;
        
      return days + 365 * y + (y + 3) / 4 - 1;
    }

    static long time2long(const uint16_t days, const uint8_t h, const uint8_t m, const uint8_t s)
    {
      return ((days * 24L + h) * 60 + m) * 60 + s;
    }
};

#endif    //  DATETIME_GENERIC_H
